arwen/host/asyncCall.go

```diff
@@ -51,21 +49,38 @@ func (host *vmHost) handleAsyncCallBreakpoint() error {
 	return nil
 }
 
-func (host *vmHost) canExecuteSynchronouslyOnDest(dest []byte) bool {
+func (host *vmHost) canExecuteSynchronously(destination []byte, data []byte) bool {
+	runtime := host.Runtime()
 	blockchain := host.Blockchain()
-	calledSCCode, err := blockchain.GetCode(dest)
 
-	return len(calledSCCode) != 0 && err == nil
-}
+	// Sync execution requires the destination to be in the same shard.
+	shardOfSC := blockchain.GetShardOfAddress(runtime.GetSCAddress())
+	shardOfDest := blockchain.GetShardOfAddress(destination)
+	if shardOfSC != shardOfDest {
+		return false
+	}
 
-func (host *vmHost) canExecuteSynchronously() bool {
-	// TODO replace with a blockchain hook that verifies if the caller and callee
-	// are in the same Shard.
-	runtime := host.Runtime()
-	asyncCallInfo := runtime.GetAsyncCallInfo()
-	dest := asyncCallInfo.Destination
+	// TODO add check for blockchain.IsPayable(destination), when it will be
+	// available.
+
+	// If ArgParser cannot read the Data field, then this isn't even a smart
+	// contract call. Async calls should not be used for transferring value.
+	argParser := parsers.NewCallArgsParser()
+	functionName, _, err := argParser.ParseData(string(data))
```

Comment:
this is not good - instead of canExecuteSynchronously it should be shouldExecuteSynchronously which should return true even if argument parser fails.

Reply:
Will rewrite.